cookbook/offscreen-clipboard-write/background.js

```diff
@@ -0,0 +1,53 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// TODO(dotproto): Add more comments explaining this code.
+
+const textToCopy = `Hello world!`;
+
+chrome.action.onClicked.addListener(async () => {
+  await addToClipboard(textToCopy);
+});
+
+// Solution 1 - Current solution for extension service workers
+
+async function addToClipboard(value) {
+  if (await chrome.offscreen.hasDocument()) {
+    console.debug('Offscreen doc already exists');
+  } else {
+    console.debug('Creating a new offscreen document.');
+
+    // TODO(dotproto): Follow up with eng to make sure that the doc's
+    // `runtime.onMessage` listener will be registered before the
+    // `createDocument()` call resolves. If not, we'll need to rewrite this to
+    // guarantee the correct order of operations.
+    await chrome.offscreen.createDocument({
+      url: 'offscreen.html',
+      reasons: [chrome.offscreen.Reason.CLIPBOARD],
+      justification: 'Write text to the clipboard.',
+    });
+  }
+
+  chrome.runtime.sendMessage({
+    type: 'copy-data-to-clipboard',
+    target: 'offscreen-doc',
+    data: value,
+  });
+}
+
+// Solution 2 – Only use this when
+
+async function addToClipboardV2(value) {
+  navigator.clipboard.copyText(value)
```

Comment:
`copyText` is not a method oc `navigator.clipboard`. 

The idea is to utilize Async Clipboard API to write a File to the clipboard with ClipboardItem and get the contents of clipboard items on arbitrary Web pages that can read from the clipboard. Hence the use of Async Clipboard API in https://github.com/GoogleChrome/chrome-extensions-samples/issues/795. This example does not achieve that.

Reply:
Thanks for pointing out the mistake. Looks like I accidentally swapped "write" for "copy". This has been addressed in the latest commit.

The original issue reported in #795 it explicitly describes writing text to and reading text from the clipboard. I'm not against adding a file demo, but that's out of scope for what we're showing in this sample.

Reply:
The original issue used _async_ Clipboard API for a reason. Perhaps I should have been very explicit about what I was trying to do.

Should I file another issue explaining precisely that the requirement is to write a `File` to the clipboard from an offscreen document - which _is_ currently impossible?

Reply:
Yep, that would be helpful for tracking and scheduling purposes. Note that creation of an extension sample for this is currently blocked by [1404835](https://bugs.chromium.org/p/chromium/issues/detail?id=1404835) (or another appropriate crbug).

Reply:
I just tried this 

```
async function writeClipImg() {
  try {
    //const imgURL = "https://avatars.githubusercontent.com/u/1047763?s=40&v=4";
    //const data = await fetch(imgURL);
    const blob = new Blob(['data'],{
      type: 'application/octet-stream'
    });

    await navigator.clipboard.write([new ClipboardItem({
      [blob.type]: blob,
    }), ]);
    console.log("Fetched image copied.");
  } catch (err) {
    console.error(err.name, err.message);
  }
}
onclick = async(e)=>{
  onclick = null;
  try {
    await writeClipImg();
    // async function getClipboardContents() {

    const clipboardItems = await navigator.clipboard.read();

    for (const clipboardItem of clipboardItems) {
      for (const type of clipboardItem.types) {
        const blob = await clipboardItem.getType(type);
        // we can now use blob here
        console.log(blob);
      }
    }

    //}
  } catch (e) {
    console.error(e);
  }
}
```

at `console` on this page using basic MDN example and got this error

```
NotAllowedError Type application/octet-stream not supported on write.
```

so appears a few things need to change for this to work as I expected. 

I already proposed this https://github.com/guest271314/requestClient which would solve https://github.com/GoogleChrome/chrome-extensions-samples/issues/766, filed as an issue here in https://github.com/GoogleChrome/chrome-extensions-samples/issues/781 - explained in the repository above. 

So, I think I have laid out several solutions for #766 already - including setting to `opener` of the offscreen document https://github.com/GoogleChrome/chrome-extensions-samples/issues/778 which was previous possible with `setSelfAsOpener` in MV2 that no longer works. Ideally though we would not load an additional document at all - just dual-opt in client and `ServiceWorker`. 

I tried to file a FUGU, however a few stakeholders have permanenetly banned me. I was able to file this a few minutes ago https://github.com/wintercg/proposal-common-minimum-api/issues/36.

> Banned from Chromium and GitHub repos of the W3C? Ooops. Then I guess you're out of luck.
>  
> 
> On Mon, Jan 9, 2023 at 3:31 AM Thomas Steiner <[tomac@google.com](mailto:tomac@google.com)> wrote:
> Hi there,
> 
> This sounds more like an (extension) service worker feature request, so the venues would be the [service worker repo](https://github.com/w3c/ServiceWorker) or the [webextensions repo](https://github.com/w3c/webextensions). You can open new issues there. 
> 
> Cheers,
> Tom

I've filed a PR, and composed two solutions for #766 (https://github.com/guest271314/sw-transfer-stream; https://github.com/guest271314/offscreen-webrtc). There is also this https://github.com/guest271314/native-messaging-quickjs/tree/dynamic-externally-connectable to dynamically set externall connectable - though we are still using JSON for communication.

When you find the time kind review the above issues and proposals. 'Bout all a guy can do at this point re #766.

Carry on.